rllib/env/multi_agent_episode.py

```diff
@@ -1286,6 +1286,192 @@ def get_truncateds(self) -> MultiAgentDict:
+        be included, but the next action and sum of rewards until this point will
+        be stored in the agent's buffer for the returned MultiAgentEpisode slice.
+
+        .. testcode::
+
+            from ray.rllib.env.multi_agent_episode import MultiAgentEpisode
+            from ray.rllib.utils.test_utils import check
+
+            # Generate a simple multi-agent episode.
+            observations = [
+                {"a0": 0, "a1": 0},  # 0
+                {         "a1": 1},  # 1
+                {         "a1": 2},  # 2
+                {"a0": 3, "a1": 3},  # 3
+                {"a0": 4},           # 4
+            ]
+            # Actions are the same as observations (except for last obs, which doesn't
+            # have an action).
+            actions = observations[:-1]
+            # Make up a reward for each action.
+            rewards = [
+                {aid: r / 10 + 0.1 for aid, r in o.items()}
+                for o in observations
+            ]
+            episode = MultiAgentEpisode(
+                observations=observations,
+                actions=actions,
+                rewards=rewards,
+                len_lookback_buffer=0,
+            )
+
+            # Slice the episode and check results.
+            slice = episode[1:3]
+            a0 = slice.agent_episodes["a0"]
+            a1 = slice.agent_episodes["a1"]
+            check((a0.observations, a1.observations), ([3], [1, 2, 3]))
+            check((a0.actions, a1.actions), ([], [1, 2]))
+            check((a0.rewards, a1.rewards), ([], [0.2, 0.3]))
+            check((a0.is_done, a1.is_done), (False, False))
+
+            # If a slice ends in a "gap" for an agent, expect actions and rewards to be
+            # cached in the agent's buffer.
+            slice = episode[:2]
+            a0 = slice.agent_episodes["a0"]
+            check(a0.observations, [0])
+            check(a0.actions, [])
+            check(a0.rewards, [])
+            check(slice._agent_buffered_actions["a0"], 0)
+            check(slice._agent_buffered_rewards["a0"], 0.1)
+
+        Args:
+            slice_: The slice object to use for slicing. This should exclude the
+                lookback buffer, which will be prepended automatically to the returned
+                slice.
+
+        Returns:
+            The new MultiAgentEpisode representing the requested slice.
+        """
+        if slice_.step not in [1, None]:
+            raise NotImplementedError(
+                "Slicing MultiAgentEnv with a step other than 1 (you used"
+                f" {slice_.step}) is not supported!"
+            )
+
+        ma_episode = MultiAgentEpisode()
+
+        slice_from_end = slice_.start is not None and slice_.start < 0
+
+        # Translate slice separately for each agent's SingleAgentEpisode, then slice
+        # those and return a new MultiAgentEpisode with the sliced SingleAgentEpisodes
+        # in it.
+        for agent_id, sa_episode in self.agent_episodes.items():
```

Comment:
Dumb questions: I see we are working here with SAEs. Couldn't we just use the MAE's getters to request all timesteps of the slice and use them in MAE.__init__ to produce a new one?

Reply:
Great point. I fixed this and cleaned up a little more (even in SingleAgentEpisode):
- SingleAgentEpisode slicing should only use `SingleAgentEpisode.get_observations/actions/..()`.
- MultiAgentEpisode slicing should work with `SingleAgentEpisode.get_observations/actions/..(return_list=True)`.